Framework/DataHandling/src/Mantid3MFFileIO.cpp

```diff
@@ -0,0 +1,512 @@
+  Lib3MF::PBuildItemIterator buildItemIterator = model->GetBuildItems();
+  while (buildItemIterator->MoveNext()) {
+    Lib3MF::PBuildItem buildItem = buildItemIterator->GetCurrent();
+    uint32_t objectResourceID = buildItem->GetObjectResourceID();
+    sLib3MFTransform transform = buildItem->GetObjectTransform();
+    readMeshObject(meshObjects, sample, objectResourceID, transform);
+  }
+}
+
+/**
+ * Attempt to read a single mesh from a specified lib3mf resource id. If the
+ * resource id points at a component instead then read them
+ * @param meshObjects A vector to store the meshes for env components
+ * @param sample A parameter to store a mesh for the sample
+ * @param objectResourceID Integer identifier of the object to read in
+ * @param transform Rotation/translation matrix for the object
+ */
+void Mantid3MFFileIO::readMeshObject(std::vector<MeshObject_sptr> &meshObjects,
+                                     MeshObject_sptr &sample,
+                                     uint32_t objectResourceID,
+                                     sLib3MFTransform transform) {
+  // no general GetObjectByID in the lib3MF library??
+  try {
+    Lib3MF::PMeshObject meshObject = model->GetMeshObjectByID(objectResourceID);
+    std::string objectName = meshObject->GetName();
+    std::transform(objectName.begin(), objectName.end(), objectName.begin(),
+                   toupper);
+    if (objectName == SAMPLE_OBJECT_NAME) {
+      sample = loadMeshObject(meshObject, transform);
+    } else {
+      meshObjects.push_back(loadMeshObject(meshObject, transform));
+    }
+  } catch (std::exception) {
+    readComponents(meshObjects, sample, objectResourceID, transform);
+  }
+}
+/*
+ * Read in the set of mesh objects pointed to by a component object
+ * @param meshObjects A vector to store the meshes for env components
+ * @param sample A parameter to store a mesh for the sample
+ * @param objectResourceID Integer identifier of the object to read in
+ * @param transform Rotation/translation matrix for the object
+ */
+void Mantid3MFFileIO::readComponents(std::vector<MeshObject_sptr> &meshObjects,
+                                     MeshObject_sptr &sample,
+                                     uint32_t objectResourceID,
+                                     sLib3MFTransform transform) {
+  Lib3MF::PComponentsObject componentsObject =
+      model->GetComponentsObjectByID(objectResourceID);
+  for (Lib3MF_uint32 nIndex = 0; nIndex < componentsObject->GetComponentCount();
+       nIndex++) {
+    Lib3MF::PComponent component = componentsObject->GetComponent(nIndex);
+    readMeshObject(meshObjects, sample, component->GetObjectResourceID(),
+                   transform);
+  }
+}
+
+/*
+ * Write a Mantid Geometry::MeshObjects to the lib3mf model object as part
+ * of processing of writing to a 3mf format file
+ * @param mantidMeshObject MeshObject to write out
+ * @param name Name of the mesh object
+ */
+void Mantid3MFFileIO::writeMeshObject(
```

Comment:
I'm guessing this is handled, but wanted to check anyway: When mesh objects are loaded the transformation matrix has to be changed due to rows being major. Does anything need to be done in that vein when writing?

Reply:
Good question - When writing out data to 3mf I don't make use of the feature in the 3mf format where you can specify object coordinates and then rotate the whole object according to a transformation matrix. I just ensure all the coordinates are as required (ie already in their final position) and specify a "do nothing" rotation matrix on L330
I think the transformation matrix feature in the 3mf format is mainly used if you have a multi-part mesh where two of the parts are copies of the same object in different positions\rotations. We don't hold this kind of information in the sample environment geometry in Mantid. Any environment containing two copies of the same shape would be stored as two separate objects with no hint that they're copies of the same thing.